io/src/real_sense/librealsense/real_sense_grabber.cpp

```diff
@@ -0,0 +1,310 @@
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include <boost/lexical_cast.hpp>
+
+#include <librealsense/rs.hpp>
+
+#include <pcl/common/io.h>
+#include <pcl/io/buffers.h>
+#include <pcl/io/real_sense_grabber.h>
+#include <pcl/io/real_sense/librealsense/real_sense_device_manager.h>
+
+
+using namespace pcl::io;
+using namespace pcl::io::real_sense;
+
+pcl::RealSenseGrabber::RealSenseGrabber (const std::string& device_id, const Mode& mode, bool strict)
+: Grabber ()
+, is_running_ (false)
+, confidence_threshold_ (6)
+, temporal_filtering_type_ (RealSense_None)
+, temporal_filtering_window_size_ (1)
+, mode_requested_ (mode)
+, strict_ (strict)
+{
+  if (device_id == "")
+    device_ = RealSenseDeviceManager::getInstance ()->captureDevice ();
+  else if (device_id[0] == '#')
+    device_ = RealSenseDeviceManager::getInstance ()->captureDevice (boost::lexical_cast<int> (device_id.substr (1)) - 1);
+  else
+    device_ = RealSenseDeviceManager::getInstance ()->captureDevice (device_id);
+
+  point_cloud_signal_ = createSignal<sig_cb_real_sense_point_cloud> ();
+  point_cloud_rgba_signal_ = createSignal<sig_cb_real_sense_point_cloud_rgba> ();
+}
+
+void
+pcl::RealSenseGrabber::start ()
+{
+  if (!is_running_)
+  {
+    need_xyz_ = num_slots<sig_cb_real_sense_point_cloud> () > 0;
+    need_xyzrgba_ = num_slots<sig_cb_real_sense_point_cloud_rgba> () > 0;
+    if (need_xyz_ || need_xyzrgba_)
+    {
+      if (temporal_filtering_type_ != RealSense_None)
+      {
+        temporal_filtering_type_ = RealSense_None;
+        PCL_WARN ("[pcl::RealSenseGrabber::enableTemporalFiltering] This device does not support temporal filter with librealsense now\n");
+      }
+      selectMode ();
+      device_->getDevice ()->enable_stream (rs::stream::depth, mode_selected_.depth_width, mode_selected_.depth_height, rs::format::z16, mode_selected_.fps);
+    }
+    if (need_xyzrgba_)
+    {
+      device_->getDevice ()->enable_stream (rs::stream::color, mode_selected_.color_width, mode_selected_.color_height, rs::format::rgb8, mode_selected_.fps);
+    }
+    frequency_.reset ();
+    is_running_ = true;
+    thread_ = boost::thread (&RealSenseGrabber::run, this);
+  }
+}
+
+const std::string&
+pcl::RealSenseGrabber::getDeviceSerialNumber () const
+{
+  return (device_->getSerialNumber ());
+}
+
+void
+pcl::RealSenseGrabber::setConfidenceThreshold (unsigned int threshold)
+{
+  if (threshold > 15)
+  {
+    PCL_WARN ("[pcl::RealSenseGrabber::setConfidenceThreshold] Attempted to set threshold outside valid range (0-15)");
+  }
+  else if (!device_->getDevice ()->supports_option (rs::option::f200_confidence_threshold))
```

Comment:
According to https://github.com/IntelRealSense/librealsense/blob/5831907ed32a52d9ff4cf7170c0580de8c283ff5/include/librealsense/rs.hpp#L61, there is only `f200_confidence_threshold` defined.